xarray/core/groupby.py

```diff
@@ -294,91 +311,282 @@ def _apply_loffset(
+            # TODO: sort instead of raising an error
+            raise ValueError("index must be monotonic for resampling")
+
+        if isinstance(group_as_index, CFTimeIndex):
+            index_grouper = CFTimeGrouper(
+                freq=grouper.freq,
+                closed=grouper.closed,
+                label=grouper.label,
+                origin=grouper.origin,
+                offset=grouper.offset,
+                loffset=grouper.loffset,
+            )
+        else:
+            index_grouper = pd.Grouper(
+                freq=grouper.freq,
+                closed=grouper.closed,
+                label=grouper.label,
+                origin=grouper.origin,
+                offset=grouper.offset,
+            )
+        self.index_grouper: CFTimeGrouper | pd.Grouper = index_grouper
+
+    def _get_index_and_items(self) -> tuple[pd.Index, pd.Series, np.ndarray]:
+        first_items, codes = self.first_items()
+        full_index = first_items.index
+        if first_items.isnull().any():
+            first_items = first_items.dropna()
+
+        full_index = full_index.rename("__resample_dim__")
+        return full_index, first_items, codes
+
+    def first_items(self) -> tuple[pd.Series, np.ndarray]:
+        from xarray import CFTimeIndex
+
+        if isinstance(self.group_as_index, CFTimeIndex):
+            return self.index_grouper.first_items(self.group_as_index)
+        else:
+            s = pd.Series(np.arange(self.group_as_index.size), self.group_as_index)
+            grouped = s.groupby(self.index_grouper)
+            first_items = grouped.first()
+            counts = grouped.count()
+            # This way we generate codes for the final output index: full_index.
+            # So for _flox_reduce we avoid one reindex and copy by avoiding
+            # _maybe_restore_empty_groups
+            codes = np.repeat(np.arange(len(first_items)), counts)
+            if self.grouper.loffset is not None:
+                _apply_loffset(self.grouper.loffset, first_items)
+            return first_items, codes
+
+    def factorize(self, squeeze: bool) -> None:
+        self.full_index, first_items, codes = self._get_index_and_items()
+        sbins = first_items.values.astype(np.int64)
+        self.group_indices = [slice(i, j) for i, j in zip(sbins[:-1], sbins[1:])] + [
+            slice(sbins[-1], None)
+        ]
+        self.unique_coord = IndexVariable(
+            self.group.name, first_items.index, self.group.attrs
+        )
+        self.codes = self.group.copy(data=codes)
+
+
+class Grouper(ABC):
+    pass
+
+
+@dataclass
+class UniqueGrouper(Grouper):
+    pass
+
+
+@dataclass(init=False)
+class BinGrouper(Grouper):
+    def __init__(self, bins, cut_kwargs: Mapping | None):
+        if duck_array_ops.isnull(bins).all():
+            raise ValueError("All bin edges are NaN.")
+
+        if cut_kwargs is None:
+            cut_kwargs = {}
+
+        self.bins = bins
+        self.cut_kwargs = cut_kwargs
```

Comment:
```suggestion
@dataclass
class BinGrouper(Grouper):
    bins: Any  # What is the typing?
    cut_kwargs: Mapping = field(default_factory=dict)

    def __post_init__(self):
        if duck_array_ops.isnull(self.bins).all():
            raise ValueError("All bin edges are NaN.")
```
`from dataclasses import dataclass, field`